2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,202 @@
+    obs_midx = lake_obs_interpolated.groupby(['date', depth_col]).temp.mean().to_frame()
+    obs_all = obs_midx.reindex(midx, fill_value=np.NaN)
+
+    # reshape to have one column per depth
+    obs_all = obs_all.unstack()
+    # flatten column levels
+    obs_all.columns = [' '.join([str(s) for s in col]).strip() for col in obs_all.columns.values]
+    # Now obs_all has dimensions of len(date_range) by len(depths)
+    # values are NaN wherever there are no observations
+    return obs_all
+
+
+def assemble_lake_data(site_id,
+                       metadata_augmented_file,
+                       obs_interpolated_file, 
+                       drivers_file,
+                       clarity_file,
+                       ice_flags_file,
+                       config):
+    """
+    Assemble raw data from one lake into equal-length sequences for
+    training/testing.
+
+    :param site_id: site_id of lake to process
+    :param metadata_augmented_file: lake metadata file, augmented with elevation
+    :param obs_interpolated_file: temperature observations file, interpolated
+        to LSTM depths
+    :param drivers_file: Meteorological drivers csv for this lake
+    :param clarity_file: Time-varying lake clarity csv
+    :param ice_flags_file: Time-varying ice flags csv for this lake
+    :param config: Snakemake config for process_mntoha
+    :returns: Numpy array of sequences with shape (# sequences,
+        sequence_length, # features + depths)
+
+    """
+    # Read hyperparameters
+    sequence_length = config['sequence_length']
+    sequence_offset = config['sequence_offset']
+    # pad time before 1st obs, e.g., 100 days
+    spinup_time = config['spinup_time'] 
+    # Discretized lake depths to use for LSTM
+    depths = np.array(config['depths'])
+
+    # Read temperature observations
+    obs_interpolated = pd.read_csv(obs_interpolated_file, parse_dates=['date'])
+    # lake observations
+    lake_obs_interpolated = obs_interpolated[obs_interpolated.site_id==site_id].copy()
+    # determine if there are any observations
+    if len(lake_obs_interpolated) == 0:
+        lake_sequences = np.empty((0))
+    else:
+        # Read metadata with elevation
+        lake_metadata_augmented = pd.read_csv(metadata_augmented_file)
+        # NOTE: If there are duplicates, the next line will only grap the first.
+        # If there are no matches, it will throw an error
+        lake = lake_metadata_augmented[lake_metadata_augmented.site_id==site_id].iloc[0,:]
+
+        # Fill NaNs for missing obs
+        obs_all = all_dates_depths(lake_obs_interpolated,
+                                   depths,
+                                   pad_before_days=spinup_time,
+                                   min_days=sequence_length)
+        # Now obs_all has dimensions of len(date_range) by len(depths)
+        # values are NaN wherever there are no observations
+
+        # Join drivers
```

Comment:
For each of these joins I'd suggest just adding a note that you're joining by date, since you're defaulting to the `join()` option to join on the index of each dataframe.